ng-dev/release/notes/changelog.ts

```diff
@@ -0,0 +1,127 @@
+import {existsSync, readFileSync, writeFileSync} from 'fs';
+import {join} from 'path';
+import * as semver from 'semver';
+import {GitClient} from '../../utils/git/git-client';
+
+/** Project-relative path for the changelog file. */
+export const changelogPath = 'CHANGELOG.md';
+
+/** Project-relative path for the changelog archive file. */
+export const changelogArchivePath = 'CHANGELOG_ARCHIVE.md';
+
+/** A marker used to split a CHANGELOG.md file into individual entries. */
+export const splitMarker = '<!-- CHANGELOG SPLIT MARKER -->';
+
+/**
+ * A string to use between each changelog entry when joining them together.
+ *
+ * Since all every changelog entry's content is trimmed, when joining back together, two new lines
+ * must be placed around the splitMarker to create a one line buffer around the comment in the
+ * markdown.
+ * i.e.
+ * <changelog entry content>
+ *
+ * <!-- CHANGELOG SPLIT MARKER -->
+ *
+ * <changelog entry content>
+ */
+const joinMarker = `\n\n${splitMarker}\n\n`;
+
+/** A RegExp matcher to extract the version of a changelog entry from the entry content. */
+const versionAnchorMatcher = new RegExp(`<a name="(.*)"></a>`);
```

Comment:
I'm a bit unhappy about the reliance on some template-specifics here. It will become problematic if we make changes of the anchor. e.g. even when we switch from `<a name` to `<a id` (which is actually better for fragment resolution in w3c).

If you see my comment on `moveEntriesPriorToVersionToArchive`, an option is that we might not need the version at all, or alternatively we could incorporate the version in a specific comment that is generated here as part of the `Changelog` class (that keeps the logic more local and self-contained)

Reply:
Should we include a metadata comment object?  Something like.

```md
<!--ReleaseMetadata {version: '1.2.3', name: 'release name'} -->
```

Then we can retrieve it with a regex and parse the results

`.  We can then `JSON.parse` the matched content into something like `Partial<ReleaseMetadata>` where we have
```ts
interface ReleaseMetadata {
  version: string;
  name: string;
}


const metadataMatcher = RegExp(/<!--ReleaseMetadata.*(\{.*\})/);

for (const entry of entries) {
  const metadata = JSON.parse(entry.match(metadataMatcher)[1]) as Partial<ReleaseMetadata>;
}
```

There would need to be more validation/assertions I suppose, but this concept.

Reply:
yeah, something around that is what I'd love to see. Could also be just something like:

```
<!-- ReleaseMetadata: Version|Name-->
```

Doesn't necessarily need to be JSON. I'm happy with either thing.

Reply:
I think it makes sense to use JSON since it will make parsing easier.  It won't be as whitespace sensitive and is easy to roll and unroll.

Reply:
👍

Reply:
Actually as I am looking at it now, I think it kind of ends up being a lot more change than we currently have to set this up, how do you feel about doing it in a follow up PR?  It just feels like it ends up nearly unrelated to the current change.

Reply:
sure, works for me!

Reply:
Creating tracking issue at: https://github.com/angular/dev-infra/issues/219